File: java/0144_preorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// 非递归
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        var res = new ArrayList<Integer>();
        var stk = new Stack<TreeNode>();
        while(!stk.empty() || root!=null)
        {
            while(root!=null)
            {
                res.add(root.val);
                stk.push(root);
                root = root.left;
            }
            root = stk.peek().right;
            stk.pop();
        }
        return res;
    }
}

//  前序遍历：根左右
// class Solution {
//     public List<Integer> preorderTraversal(TreeNode root) {
//         var res = new ArrayList<Integer>();
//         traverse(res,root);
//         return res;
//     }
//     public void traverse(List<Integer> res, TreeNode root)
//     {
//         if(root==null)
//             return;
//         res.add(root.val);
//         traverse(res,root.left);
//         traverse(res,root.right);
//     }
// }

File: cpp/0144_preorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// 非递归：栈
class Solution {
public:

    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> res;
        // stack 的作用，每个元素都是下一个元素的根节点
        stack<TreeNode*> stk;
        while(!stk.empty() || root!=nullptr)
        {
            while(root!=nullptr)
            {
                res.push_back(root->val);
                stk.push(root);
                root = root->left;
            }
            root = stk.top();
            stk.pop();
            root = root->right;
        }
        return res;
    }
};

// // 前序遍历：根左右
// class Solution {
// public:

//     vector<int> preorderTraversal(TreeNode* root) {
//         vector<int> res;
//         traverse(res,root);
//         return res;
//     }

//     void traverse(vector<int> &res, TreeNode *root)
//     {
//         if(root == nullptr)
//             return;
        
//         res.push_back(root->val);
//         traverse(res,root->left);
//         traverse(res,root->right);
//     }
// };
